src/Grids/orthogonal_spherical_shell_grid.jl

```diff
@@ -801,15 +801,76 @@ function Base.show(io::IO, grid::OrthogonalSphericalShellGrid, withsummary=true)
+@inline xnodes(grid::OSSG, LX::Face,   LY::Center; with_halos=false) = with_halos ? grid.λᶠᶜᵃ :
+    view(grid.λᶠᶜᵃ, interior_indices(typeof(LX), topology(grid, 1), grid.Nx), interior_indices(typeof(LY), topology(grid, 2), grid.Ny))
+@inline xnodes(grid::OSSG, LX::Center, LY::Face, ; with_halos=false) = with_halos ? grid.λᶜᶠᵃ :
+    view(grid.λᶜᶠᵃ, interior_indices(typeof(LX), topology(grid, 1), grid.Nx), interior_indices(typeof(LY), topology(grid, 2), grid.Ny))
+@inline xnodes(grid::OSSG, LX::Center, LY::Center; with_halos=false) = with_halos ? grid.λᶜᶜᵃ :
+    view(grid.λᶜᶜᵃ, interior_indices(typeof(LX), topology(grid, 1), grid.Nx), interior_indices(typeof(LY), topology(grid, 2), grid.Ny))
+
+@inline ynodes(grid::OSSG, LX::Face,   LY::Face, ; with_halos=false) = with_halos ? grid.φᶠᶠᵃ :
+    view(grid.φᶠᶠᵃ, interior_indices(typeof(LX), topology(grid, 1), grid.Nx), interior_indices(typeof(LY), topology(grid, 2), grid.Ny))
+@inline ynodes(grid::OSSG, LX::Face,   LY::Center; with_halos=false) = with_halos ? grid.φᶠᶜᵃ :
+    view(grid.φᶠᶜᵃ, interior_indices(typeof(LX), topology(grid, 1), grid.Nx), interior_indices(typeof(LY), topology(grid, 2), grid.Ny))
+@inline ynodes(grid::OSSG, LX::Center, LY::Face, ; with_halos=false) = with_halos ? grid.φᶜᶠᵃ :
+    view(grid.φᶜᶠᵃ, interior_indices(typeof(LX), topology(grid, 1), grid.Nx), interior_indices(typeof(LY), topology(grid, 2), grid.Ny))
+@inline ynodes(grid::OSSG, LX::Center, LY::Center; with_halos=false) = with_halos ? grid.φᶜᶜᵃ :
+    view(grid.φᶜᶜᵃ, interior_indices(typeof(LX), topology(grid, 1), grid.Nx), interior_indices(typeof(LY), topology(grid, 2), grid.Ny))
+
+@inline xnodes(grid::OSSG, LX::CellLocation, LY::CellLocation, LZ::CellLocation; kwargs...) = xnodes(grid, LX, LY; kwargs...)
+@inline ynodes(grid::OSSG, LX::CellLocation, LY::CellLocation, LZ::CellLocation; kwargs...) = ynodes(grid, LX, LY; kwargs...)
+
+@inline znodes(grid::OSSG, LZ::Face  ; with_halos=false) = with_halos ? grid.zᵃᵃᶠ : view(grid.zᵃᵃᶠ, interior_indices(typeof(LZ), topology(grid, 3), grid.Nz))
+@inline znodes(grid::OSSG, LZ::Center; with_halos=false) = with_halos ? grid.zᵃᵃᶜ : view(grid.zᵃᵃᶜ, interior_indices(typeof(LZ), topology(grid, 3), grid.Nz))
+
+
+@inline xnode(i, j, grid::OSSG, LX::CellLocation, LY::CellLocation) = xnodes(grid, LX, LY, with_halos=true)[i, j]
+@inline ynode(i, j, grid::OSSG, LX::CellLocation, LY::CellLocation) = ynodes(grid, LX, LY, with_halos=true)[i, j]
+@inline znode(k, grid::OSSG, LZ::CellLocation)                      = znodes(grid, LZ, with_halos=true)[k]
+
+@inline xnode(i, j, k, grid::OSSG, LX::CellLocation, LY::CellLocation, LZ::CellLocation) = xnode(i, j, grid, LX, LY)
+@inline ynode(i, j, k, grid::OSSG, LX::CellLocation, LY::CellLocation, LZ::CellLocation) = ynode(i, j, grid, LY, LY)
+@inline znode(i, j, k, grid::OSSG, LX::CellLocation, LY::CellLocation, LZ::CellLocation) = znode(k, grid, LZ)
```

Comment:
I'm a bit confused, e.g., with:

```
@inline znode(k, grid::OSSG, LZ::CellLocation)                      = znodes(grid, LZ, with_halos=true)[k]
@inline znode(i, j, k, grid::OSSG, LX::CellLocation, LY::CellLocation, LZ::CellLocation) = znode(k, grid, LZ)
```

So the second one calls the first and that calls `znodes()` which computes all the nodes only to select then the `k`-th one? Isn't this a bit of a waste?

Reply:
Yes, that's what it is. It's by design since if we change the name of `Δzᵃᵃᶠ` we just need to do it once in `znodes()`. So my idea was to help code maintainability.

Plus, if I understand it correctly, the spacings `grid.Δzᵃᵃᶠ` etc. for all grids are calculated once and stored in `grid`, so we're not actually computing them each time we call `zspacing()`.

Reply:
I'm sorry but I don't understand.

What was the problem with the previous way `znode()` was implemented?

Perhaps I got into this discussion late and that's why I don't understand...?

Reply:
It may also just be a bad decision on my part. But just to explain, I'll take `z`-nodes on a `Center` for a `RectilinearGrid` as an example.  Currently the nodes functions are defined here:

https://github.com/CliMA/Oceananigans.jl/blob/09ff3a9865fefde7921c63321073c61406f64bb5/src/Grids/rectilinear_grid.jl#L364-L376

The old `all_z_nodes(Center, grid)` is what's now `znodes(grid, Center())`.

At some point (I think on https://github.com/CliMA/Oceananigans.jl/issues/2058) @glwagner suggested that we should have one method for `znode()` and `znodes()` that take all three cell locations (i.e. `LX`, `LY`, `LZ`; to maintain some uniformity) and whenever possible that we should have another method that takes only the necessary location for the `z` position to be retrieved (`LZ` for a rect grid). So to implement that we need the methods

```julia
znodes(grid, LZ)
znodes(grid, LX, LY, LZ)
znode(k, grid, LZ)
znode(i, j, k, grid, LX, LY, LZ)
```

We could implement them as something like

```julia
znodes(grid, LZ)  = grid.zᵃᵃᶠ
znodes(grid, LX, LY, LZ)  = grid.zᵃᵃᶠ
znode(k, grid, LZ) =  grid.zᵃᵃᶠ[k]
znode(i, j, k, grid, LX, LY, LZ) = grid.zᵃᵃᶠ[k]
```

But my thinking was that, apart from being repetitive, if we decide to change the name of  `zᵃᵃᶠ` to something else in the future, we need to change that in at least 4 places. Which is why I chose to implement this as

```julia
znodes(grid, LZ)  = grid.zᵃᵃᶠ
znodes(grid, LX, LY, LZ)  = znodes(grid, LZ)
znode(k, grid, LZ) =  znodes(grid, LZ)[k]
znode(i, j, k, grid, LX, LY, LZ) = znode(k, grid, LZ)
```

That way the broader functions build on top of the more specialized ones. It's a bit more complicated for the spacing functions (since they can be a `Number` or an `Array` depending if the direction is stretched or not), but the same reasoning goes. 

We can definitely change how these are implemented in this PR, but that was my rationale for doing it the way I did. @navidcy I hope this helps you understand better! Let me know what needs more clarification.

Reply:
Why not something like:

```Julia
znodes(grid, ::Face)  = grid.zᵃᵃᶠ
znodes(grid, ::Center)  = grid.zᵃᵃᶜ
znodes(grid, LX, LY, LZ)  = znodes(grid, LZ)

znode(k, grid, LZ) =  grid.zᵃᵃᶠ[k]
znode(i, j, k, grid, LX, LY, LZ) = znode(k, grid, LZ) 
```

Reply:
Yeah I can do that too, np. Does this have advantages over the way it's currently written?

Reply:
it reads more natural in my head... 

I feel bad making suggestions if things have been discussed already.

@glwagner you have an opinion on this? it's just that e.g.
```Julia
znode(k, grid, LZ) = znodes(grid, LZ, with_halos=true)[k]
```
feels weird --- is it just me?

Reply:
> I feel bad making suggestions if things have been discussed already.

These minutiae weren't discussed. Just the methods. So you're in a good spot for making suggestions!

Reply:
@tomchor mind if I push a few things?

I started having to add some of this functionality in #2867 so it'd be better to finish this otherwise it'd be double work. ;)

Reply:
Go for it!